File: sample/Farkle.Samples.FSharp/JSON.fs
// Copyright (c) 2019 Theodore Tsirpanis
//
// This software is released under the MIT License.
// https://opensource.org/licenses/MIT

module Farkle.Samples.FSharp.JSON

open System
open System.Globalization
open System.Text.Json.Nodes
open Farkle.Builder

let toDecimal (x: ReadOnlySpan<char>) =
    Decimal.Parse(
    #if NETCOREAPP
        x,
    #else
        x.ToString(),
    #endif
        NumberStyles.Float, CultureInfo.InvariantCulture)
        |> JsonValue.Create
        :> JsonNode

open Regex

let builder =
    // Better let that regex stay.
    // JSON prohibits leading zeroes,
    // and we want to avoid boxing.
    let number =
        let numberChars = charRanges ['0', '9']
        concat [
            char '-' |> optional
            choice [
                char '0'
                chars "123456789" + (numberChars |> star)
            ]
            optional <| (char '.' + (numberChars |> atLeast 1))
            [chars "eE"; chars "+-" |> optional; numberChars |> atLeast 1]
            |> concat
            |> optional
        ]
        |> terminal "Number" (T(fun _ data -> toDecimal data))
    let string = Terminals.stringEx "/bfnrtu" false '"' "String"
    let object = nonterminal "Object"
    let array = nonterminal "Array"
    let value = "Value" ||= [
        !@ string => (fun str -> JsonValue.Create str :> JsonNode)
        !@ number |> asProduction
        !@ object |> asProduction
        !@ array |> asProduction
        !& "true" => (fun () -> JsonValue.Create true :> JsonNode)
        !& "false" => (fun () -> JsonValue.Create false :> JsonNode)
        !& "null" =% null
    ]
    let arrayReversed: Nonterminal<JsonArray> = nonterminal "Array Reversed"
    arrayReversed.SetProductions(
        !@ arrayReversed .>> "," .>>. value => (fun xs x -> xs.Add x; xs),
        !@ value => (fun x -> let xs = JsonArray() in xs.Add(x); xs)
    )
    let arrayOptional = "Array Optional" ||= [
        !@ arrayReversed |> asProduction
        empty => (fun () -> JsonArray())
    ]
    array.SetProductions(!& "[" .>>. arrayOptional .>> "]" => (fun x -> x :> JsonNode))

    let objectElement: Nonterminal<JsonObject> = nonterminal "Object Element"
    objectElement.SetProductions(
        !@ objectElement .>> "," .>>. string .>> ":" .>>. value => (fun xs k v -> xs.Add(k, v); xs),
        !@ string .>> ":" .>>. value => (fun k v -> let obj = JsonObject() in obj.Add(k, v); obj)
    )
    let objectOptional = "Object Optional" ||= [
        !@ objectElement |> asProduction
        empty => (fun () -> JsonObject())
    ]
    object.SetProductions(!& "{" .>>. objectOptional .>> "}" => (fun x -> x :> JsonNode))

    value
    |> _.CaseSensitive(true)

let parser = GrammarBuilder.build builder

File: sample/Farkle.Samples.CSharp/JSON.cs
// Copyright © Theodore Tsirpanis and Contributors.
// SPDX-License-Identifier: MIT

using Farkle.Builder;
using static Farkle.Builder.Regex;
using System;
using System.Globalization;
using System.Text.Json.Nodes;
using System.Collections.Immutable;
using Farkle.Parser;

// ReSharper disable once CheckNamespace
namespace Farkle.Samples.CSharp
{
    public static class JSON
    {
        private static JsonValue ToDecimal(ref ParserState state, ReadOnlySpan<char> data)
        {
            var num =
                decimal.Parse(data, NumberStyles.AllowExponent | NumberStyles.Float, CultureInfo.InvariantCulture);
            return JsonValue.Create(num);
        }

        public static readonly IGrammarBuilder<JsonNode?> Builder;

        public static readonly CharParser<JsonNode?> Parser;

        [UseEnhancedSyntax]
        static JSON()
        {
            var number = Terminal.Create("Number",
                Join(
                    Literal('-').Optional(),
                    Literal('0') | OneOf("123456789") + OneOf("0123456789").ZeroOrMore(),
                    (Literal('.') + OneOf("0123456789").AtLeast(1)).Optional(),
                    Join(
                        OneOf("eE"),
                        OneOf("+-").Optional(),
                        OneOf("0123456789").AtLeast(1)).Optional()),
                ToDecimal);
            var jsonString = Terminals.String("String", '"', "/bfnrtu", false);
            var jsonObject = Nonterminal.Create<JsonObject>("Object");
            var jsonArray = Nonterminal.Create<JsonArray>("Array");
            var value = Nonterminal.Create("Value",
                jsonString.Finish(x => JsonValue.Create(x)),
                number.AsProduction(),
                jsonObject.AsProduction(),
                jsonArray.AsProduction(),
                "true".Finish(() => JsonValue.Create(true)),
                "false".Finish(() => JsonValue.Create(false)),
                "null".FinishConstant((JsonNode?)null));

            var arrayReversed = Nonterminal.Create<JsonArray>("Array Reversed");
            arrayReversed.SetProductions(
                Production.Create(arrayReversed, ",", value).Finish((xs, x) =>
                {
                    xs.Add(x);
                    return xs;
                }),
                value.Finish(x => new JsonArray { x }));
            var arrayOptional = Nonterminal.Create("Array Optional",
                arrayReversed.AsProduction(),
                ProductionBuilder.Empty.Finish(() => new JsonArray()));
            jsonArray.SetProductions(Production.Create("[", arrayOptional, "]").AsProduction());

            var objectElement = Nonterminal.Create<JsonObject>("Object Element");
            objectElement.SetProductions(
                Production.Create(objectElement, ",", jsonString, ":", value)
                    .Finish((xs, k, v) =>
                    {
                        xs.Add(k, v);
                        return xs;
                    }),
                Production.Create(jsonString, ":", value)
                    .Finish((k, v) => new JsonObject { { k, v } }));
            var objectOptional = Nonterminal.Create("Object Optional",
                objectElement.AsProduction(),
                ProductionBuilder.Empty.Finish(() => new JsonObject()));
            jsonObject.SetProductions(Production.Create("{", objectOptional, "}").AsProduction());

            Builder = value.CaseSensitive();
            Parser = Builder.Build();

            static Regex OneOf(string chars) => Regex.OneOf(chars.ToImmutableArray());
        }
    }
}
